File: app/src/main/java/com/mmjang/ankihelper/ui/plan/PlansManagerActivity.java
package com.mmjang.ankihelper.ui.plan;

import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.provider.ContactsContract;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import androidx.core.app.NavUtils;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.ItemTouchHelper;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.Toast;

import com.mmjang.ankihelper.MyApplication;
import com.mmjang.ankihelper.R;
import com.mmjang.ankihelper.data.Settings;
import com.mmjang.ankihelper.data.database.AppDatabase;
import com.mmjang.ankihelper.data.plan.OutputPlanEntity;
import com.mmjang.ankihelper.data.plan.OutputPlanPOJO;
import com.mmjang.ankihelper.data.plan.OutputPlanRepository;
import com.mmjang.ankihelper.data.plan.OutputPlanRepositoryHelper;
import com.mmjang.ankihelper.ui.plan.helper.SimpleItemTouchHelperCallback;
import com.mmjang.ankihelper.util.DialogUtil;
import com.mmjang.ankihelper.util.Utils;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

public class PlansManagerActivity extends AppCompatActivity {

    private List<OutputPlanPOJO> mPlanList;
    RecyclerView planListView;
    PlansAdapter mPlansAdapter;
    private static final String PLAN_SEP = "|||";
    private static final int ERROR_FORMAT = 1;
    private OutputPlanRepositoryHelper planRepositoryHelper;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        if(Settings.getInstance(this).getPinkThemeQ()){
            setTheme(R.style.AppThemePink);
        }
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_plans_manager);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        // Initialize repository helper
        AppDatabase database = AppDatabase.Companion.getInstance(getApplicationContext());
        OutputPlanRepository repository = new OutputPlanRepository(database.outputPlanDao());
        planRepositoryHelper = new OutputPlanRepositoryHelper(repository, this);

        // Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        FloatingActionButton fab = (FloatingActionButton) findViewById(R.id.add_plan);
        fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (MyApplication.getAnkiDroid(getApplicationContext()).isAnkiDroidRunning()) {
                    Intent intent = new Intent(PlansManagerActivity.this, PlanEditorActivity.class);
                    startActivity(intent);
                } else {
                    DialogUtil.showStartAnkiDialog(PlansManagerActivity.this);
                }
            }
        });
        initPlanList();

        new Thread(
                new Runnable() {
                    @Override
                    public void run() {
                        try{
                            MyApplication.getAnkiDroid(getApplicationContext()).getApi().getDeckList();
                        }catch (Exception e){
                        }
                    }
                }
        ).start();
    }

    @Override
    protected void onResume() {
        super.onResume();
        // Load plans using repository helper
        planRepositoryHelper.getAllPlans(new OutputPlanRepositoryHelper.PlansCallback() {
            @Override
            public void onSuccess(List<OutputPlanEntity> entities) {
                // Convert entities to POJOs
                List<OutputPlanPOJO> newList = convertEntitiesToPOJOs(entities);
                mPlanList.clear();
                mPlanList.addAll(newList);
                mPlansAdapter.notifyDataSetChanged();
            }

            @Override
            public void onError(Throwable error) {
                Toast.makeText(PlansManagerActivity.this,
                        "Failed to load plans: " + error.getMessage(),
                        Toast.LENGTH_SHORT).show();
            }
        });
    }

    private void initPlanList() {
        mPlanList = new ArrayList<>();
        //Log.d("PlansManager:", plans.size() + "ge");
        planListView = (RecyclerView) findViewById(R.id.plan_list);
        LinearLayoutManager llm = new LinearLayoutManager(this);
        planListView.setLayoutManager(llm);
        mPlansAdapter = new PlansAdapter(PlansManagerActivity.this, mPlanList);
        //planList.addItemDecoration(new DividerItemDecoration(this, LinearLayoutManager.VERTICAL));
        planListView.setAdapter(mPlansAdapter);

        ItemTouchHelper.Callback callback = new SimpleItemTouchHelperCallback(mPlansAdapter);
        ItemTouchHelper itemTouchHelper = new ItemTouchHelper(callback);
        itemTouchHelper.attachToRecyclerView(planListView);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.activity_plans_manager_menu, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {

      if (item.getItemId() ==R.id.menu_item_export_plan) {
        exportPlans();
      }else if (item.getItemId() == R.id.menu_item_import_plan) {
        importPlans();
      } else if (item.getItemId() ==android.R.id.home) {
        NavUtils.navigateUpFromSameTask(this);
      }

        return true;
    }

    private void importPlans() {
        ClipboardManager clipboard = (ClipboardManager) getSystemService(Context.CLIPBOARD_SERVICE);
        if(clipboard.hasPrimaryClip()){
            if(clipboard.getText()!=null){
                String plansString = clipboard.getText().toString();
                processPlanString(plansString);
            }
        }else{
            Toast.makeText(this, "剪贴板为空！", Toast.LENGTH_SHORT).show();
        }
    }

    private void processPlanString(String plansString) {
        String[] lines = plansString.split("\n");
        if(lines.length == 0){
            Toast.makeText(this, "格式错误！", Toast.LENGTH_SHORT).show();
            return ;
        }

        for(String line : lines){
            if(line.replace(" ","").replace("\t", "").equals("")){
                continue;//blank line
            }
            String[] items = line.split("\\|\\|\\|");
            if(items.length != 5){
                String errorMessage = line;
                errorMessage += "\n格式错误，每行项目数应为5";
                Toast.makeText(this, errorMessage, Toast.LENGTH_SHORT).show();
                continue;
            }
            try {
                String planName = items[0].trim();
                long deckId = Long.parseLong(items[1]);
                long modeld = Long.parseLong(items[2]);
                String dictKey = items[3].trim();
                String fieldMapString = items[4];
                for(OutputPlanPOJO outputPlan : mPlanList){
                    if(outputPlan.getPlanName().equals(planName)){
                        planName = planName + "_copy";
                        break;
                    }
                }
                OutputPlanEntity outputPlanEntity = new OutputPlanEntity();
                outputPlanEntity.setPlanName(planName);
                outputPlanEntity.setOutputDeckId(deckId);
                outputPlanEntity.setOutputModelId(modeld);
                outputPlanEntity.setDictionaryKey(dictKey);
                outputPlanEntity.setFieldsMap(fieldMapString);

                planRepositoryHelper.savePlan(outputPlanEntity, new OutputPlanRepositoryHelper.OperationCallback() {
                    @Override
                    public void onSuccess() {
                        // Plan saved successfully
                    }

                    @Override
                    public void onError(Throwable error) {
                        Toast.makeText(PlansManagerActivity.this,
                                "Failed to save plan: " + error.getMessage(),
                                Toast.LENGTH_SHORT).show();
                    }
                });
            }
            catch (Exception e){
                Toast.makeText(this, e.getLocalizedMessage(), Toast.LENGTH_SHORT).show();
            }
        }
        onResume();
    }

    private List<OutputPlanPOJO> convertEntitiesToPOJOs(List<OutputPlanEntity> entities) {
        List<OutputPlanPOJO> pojos = new ArrayList<>();
        for (OutputPlanEntity entity : entities) {
            OutputPlanPOJO pojo = new OutputPlanPOJO();
            pojo.setPlanName(entity.getPlanName());
            pojo.setDictionaryKey(entity.getDictionaryKey());
            pojo.setOutputDeckId(entity.getOutputDeckId());
            pojo.setOutputModelId(entity.getOutputModelId());
            pojo.setFieldsMapString(entity.getFieldsMap());
            pojos.add(pojo);
        }
        return pojos;
    }

    private OutputPlanEntity convertPOJOToEntity(OutputPlanPOJO pojo) {
        OutputPlanEntity entity = new OutputPlanEntity();
        entity.setPlanName(pojo.getPlanName());
        entity.setDictionaryKey(pojo.getDictionaryKey());
        entity.setOutputDeckId(pojo.getOutputDeckId());
        entity.setOutputModelId(pojo.getOutputModelId());
        entity.setFieldsMap(pojo.getFieldsMapString());
        return entity;
    }

    private void exportPlans() {
        StringBuilder sb = new StringBuilder();
        for(OutputPlanPOJO plan : mPlanList){
            sb.append(plan.getPlanName());
            sb.append(PLAN_SEP);
            sb.append(plan.getOutputDeckId());
            sb.append(PLAN_SEP);
            sb.append(plan.getOutputModelId());
            sb.append(PLAN_SEP);
            sb.append(plan.getDictionaryKey());
            sb.append(PLAN_SEP);
            sb.append(plan.getFieldsMapString());
            sb.append("\n");
        }
        String exportedString = sb.toString();
        ClipboardManager clipboard = (ClipboardManager) getSystemService(Context.CLIPBOARD_SERVICE);
        ClipData clip = ClipData.newPlainText("plans string", exportedString);
        clipboard.setPrimaryClip(clip);
    }

}

File: app/src/main/java/com/lmyby/ankiquicker/ui/plan/PlansManagerActivity.kt
package com.lmyby.ankiquicker.ui.plan

import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.NavUtils
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.lmyby.ankiquicker.MyApplication
import com.lmyby.ankiquicker.R
import com.lmyby.ankiquicker.data.Settings
import com.lmyby.ankiquicker.data.database.AppDatabase
import com.lmyby.ankiquicker.data.plan.OutputPlanEntity
import com.lmyby.ankiquicker.data.plan.OutputPlanPOJO
import com.lmyby.ankiquicker.data.plan.OutputPlanRepository
import com.lmyby.ankiquicker.data.plan.OutputPlanRepositoryHelper
import com.lmyby.ankiquicker.ui.plan.helper.SimpleItemTouchHelperCallback
import com.lmyby.ankiquicker.util.DialogUtil

class PlansManagerActivity : AppCompatActivity() {

    private lateinit var mPlanList: MutableList<OutputPlanPOJO>
    private lateinit var planListView: RecyclerView
    private lateinit var mPlansAdapter: PlansAdapter
    private lateinit var planRepositoryHelper: OutputPlanRepositoryHelper

    override fun onCreate(savedInstanceState: Bundle?) {
        if (Settings.getInstance(this).getPinkThemeQ()) {
            setTheme(R.style.AppThemePink)
        }
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_plans_manager)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        // Initialize repository helper
        val database = AppDatabase.getInstance(applicationContext)
        val repository = OutputPlanRepository(database.outputPlanDao())
        planRepositoryHelper = OutputPlanRepositoryHelper(repository, this)

        val fab = findViewById<FloatingActionButton>(R.id.add_plan)
        fab.setOnClickListener {
            if (MyApplication.getAnkiDroid(applicationContext).isAnkiDroidRunning) {
                val intent = Intent(this@PlansManagerActivity, PlanEditorActivity::class.java)
                startActivity(intent)
            } else {
                DialogUtil.showStartAnkiDialog(this@PlansManagerActivity)
            }
        }

        initPlanList()

        Thread {
            try {
                MyApplication.getAnkiDroid(applicationContext).api.getDeckList()
            } catch (e: Exception) {
                // Silently fail
            }
        }.start()
    }

    override fun onResume() {
        super.onResume()
        // Load plans using repository helper
        planRepositoryHelper.getAllPlans(object : OutputPlanRepositoryHelper.PlansCallback {
            override fun onSuccess(plans: MutableList<OutputPlanEntity>) {
                // Convert entities to POJOs
                val newList = convertEntitiesToPOJOs(plans)
                mPlanList.clear()
                mPlanList.addAll(newList)
                mPlansAdapter.notifyDataSetChanged()
            }

            override fun onError(error: Throwable) {
                Toast.makeText(
                    this@PlansManagerActivity,
                    "Failed to load plans: ${error.message}",
                    Toast.LENGTH_SHORT
                ).show()
            }
        })
    }

    private fun initPlanList() {
        mPlanList = ArrayList()
        planListView = findViewById(R.id.plan_list)
        val llm = LinearLayoutManager(this)
        planListView.layoutManager = llm
        mPlansAdapter = PlansAdapter(this@PlansManagerActivity, mPlanList)
        planListView.adapter = mPlansAdapter

        val callback: ItemTouchHelper.Callback = SimpleItemTouchHelperCallback(mPlansAdapter)
        val itemTouchHelper = ItemTouchHelper(callback)
        itemTouchHelper.attachToRecyclerView(planListView)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater: MenuInflater = menuInflater
        inflater.inflate(R.menu.activity_plans_manager_menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.menu_item_export_plan -> exportPlans()
            R.id.menu_item_import_plan -> importPlans()
            android.R.id.home -> NavUtils.navigateUpFromSameTask(this)
        }
        return true
    }

    private fun importPlans() {
        val clipboard = getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
        if (clipboard.hasPrimaryClip()) {
            clipboard.text?.let { text ->
                val plansString = text.toString()
                processPlanString(plansString)
            }
        } else {
            Toast.makeText(this, "剪贴板为空！", Toast.LENGTH_SHORT).show()
        }
    }

    private fun processPlanString(plansString: String) {
        val lines = plansString.split("\n")
        if (lines.isEmpty()) {
            Toast.makeText(this, "格式错误！", Toast.LENGTH_SHORT).show()
            return
        }

        for (line in lines) {
            if (line.replace(" ", "").replace("\t", "").isEmpty()) {
                continue // blank line
            }

            val items = line.split("\\|\\|\\|".toRegex())
            if (items.size != 5) {
                val errorMessage = "$line\n格式错误，每行项目数应为5"
                Toast.makeText(this, errorMessage, Toast.LENGTH_SHORT).show()
                continue
            }

            try {
                var planName = items[0].trim()
                val deckId = items[1].toLong()
                val modelId = items[2].toLong()
                val dictKey = items[3].trim()
                val fieldMapString = items[4]

                // Check for duplicate plan names
                for (outputPlan in mPlanList) {
                    if (outputPlan.planName == planName) {
                        planName = "${planName}_copy"
                        break
                    }
                }

                val outputPlanEntity = OutputPlanEntity(
                    planName = planName,
                    dictionaryKey = dictKey,
                    outputDeckId = deckId,
                    outputModelId = modelId,
                    fieldsMap = fieldMapString
                )

                planRepositoryHelper.savePlan(outputPlanEntity, object : OutputPlanRepositoryHelper.OperationCallback {
                    override fun onSuccess() {
                        // Plan saved successfully
                    }

                    override fun onError(error: Throwable) {
                        Toast.makeText(
                            this@PlansManagerActivity,
                            "Failed to save plan: ${error.message}",
                            Toast.LENGTH_SHORT
                        ).show()
                    }
                })
            } catch (e: Exception) {
                Toast.makeText(this, e.localizedMessage, Toast.LENGTH_SHORT).show()
            }
        }
        onResume()
    }

    private fun convertEntitiesToPOJOs(entities: List<OutputPlanEntity>): List<OutputPlanPOJO> {
        val pojos = ArrayList<OutputPlanPOJO>()
        for (entity in entities) {
            val pojo = OutputPlanPOJO().apply {
                planName = entity.planName ?: ""
                dictionaryKey = entity.dictionaryKey ?: ""
                outputDeckId = entity.outputDeckId
                outputModelId = entity.outputModelId
                setFieldsMapString(entity.fieldsMap ?: "")
            }
            pojos.add(pojo)
        }
        return pojos
    }

    private fun convertPOJOToEntity(pojo: OutputPlanPOJO): OutputPlanEntity {
        return OutputPlanEntity(
            planName = pojo.planName,
            dictionaryKey = pojo.dictionaryKey,
            outputDeckId = pojo.outputDeckId,
            outputModelId = pojo.outputModelId,
            fieldsMap = pojo.getFieldsMapString()
        )
    }

    private fun exportPlans() {
        val sb = StringBuilder()
        for (plan in mPlanList) {
            sb.append(plan.planName)
            sb.append(PLAN_SEP)
            sb.append(plan.outputDeckId)
            sb.append(PLAN_SEP)
            sb.append(plan.outputModelId)
            sb.append(PLAN_SEP)
            sb.append(plan.dictionaryKey)
            sb.append(PLAN_SEP)
            sb.append(plan.getFieldsMapString())
            sb.append("\n")
        }
        val exportedString = sb.toString()
        val clipboard = getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
        val clip = ClipData.newPlainText("plans string", exportedString)
        clipboard.setPrimaryClip(clip)
    }

    companion object {
        private const val PLAN_SEP = "|||"
        private const val ERROR_FORMAT = 1
    }
}
